File: python/search_schools.py
"""
Search schools using the SchoolDigger API.

Searches for schools in a given state matching a keyword and displays
results in a formatted table.

Usage:
    export SCHOOLDIGGER_APP_ID=your_app_id
    export SCHOOLDIGGER_APP_KEY=your_app_key
    python search_schools.py
"""

import os
import sys
import requests


API_BASE = "https://api.schooldigger.com/v2.3"


def get_credentials() -> tuple[str, str]:
    """Read API credentials from environment variables."""
    app_id = os.environ.get("SCHOOLDIGGER_APP_ID")
    app_key = os.environ.get("SCHOOLDIGGER_APP_KEY")
    if not app_id or not app_key:
        print("Error: Set SCHOOLDIGGER_APP_ID and SCHOOLDIGGER_APP_KEY environment variables.")
        print("Get a free API key at https://developer.schooldigger.com")
        sys.exit(1)
    return app_id, app_key


def search_schools(state: str, query: str, app_id: str, app_key: str) -> dict:
    """Search schools by state and keyword."""
    url = f"{API_BASE}/schools"
    params = {
        "st": state,
        "q": query,
        "appID": app_id,
        "appKey": app_key,
    }
    response = requests.get(url, params=params)
    response.raise_for_status()
    return response.json()


def display_results(data: dict) -> None:
    """Print search results as a formatted table."""
    schools = data.get("schoolList", [])
    total = data.get("numberOfSchools", 0)

    if not schools:
        print("No schools found.")
        return

    print(f"\nFound {total} school(s). Showing first {len(schools)}:\n")

    # Table header
    header = f"{'School Name':<40} {'City':<18} {'Grades':<10} {'Rank':<14} {'Stars':<6} {'Students':<8}"
    print(header)
    print("-" * len(header))

    for school in schools:
        name = school.get("schoolName", "N/A")[:39]
        city = school.get("address", {}).get("city", "N/A")[:17]
        low = school.get("lowGrade", "?")
        high = school.get("highGrade", "?")
        grades = f"{low}-{high}"

        # Ranking info (most recent year)
        rank_history = school.get("rankHistory", [])
        if rank_history:
            r = rank_history[0]
            rank = f"{r.get('rank', '?')} of {r.get('rankOf', '?')}"
            stars = f"{r.get('rankStars', 'N/A')}"
        else:
            rank = "N/A"
            stars = "N/A"

        # Enrollment (most recent year)
        yearly = school.get("schoolYearlyDetails", [])
        students = str(yearly[0].get("numberOfStudents", "N/A")) if yearly else "N/A"

        print(f"{name:<40} {city:<18} {grades:<10} {rank:<14} {stars:<6} {students:<8}")


def main() -> None:
    app_id, app_key = get_credentials()

    state = "WA"
    query = "Lincoln"
    print(f"Searching for '{query}' schools in {state}...")

    try:
        data = search_schools(state, query, app_id, app_key)
        display_results(data)
    except requests.exceptions.HTTPError as e:
        print(f"API error: {e.response.status_code} — {e.response.text}")
        sys.exit(1)
    except requests.exceptions.RequestException as e:
        print(f"Network error: {e}")
        sys.exit(1)


if __name__ == "__main__":
    main()

File: javascript/search_schools.mjs
/**
 * Search schools using the SchoolDigger API.
 *
 * Searches for schools in a given state matching a keyword and displays
 * results in a formatted table.
 *
 * Usage:
 *   export SCHOOLDIGGER_APP_ID=your_app_id
 *   export SCHOOLDIGGER_APP_KEY=your_app_key
 *   node search_schools.mjs
 *
 * Requires Node.js 18+ (uses native fetch).
 */

const API_BASE = "https://api.schooldigger.com/v2.3";

function getCredentials() {
  const appID = process.env.SCHOOLDIGGER_APP_ID;
  const appKey = process.env.SCHOOLDIGGER_APP_KEY;
  if (!appID || !appKey) {
    console.error("Error: Set SCHOOLDIGGER_APP_ID and SCHOOLDIGGER_APP_KEY environment variables.");
    console.error("Get a free API key at https://developer.schooldigger.com");
    process.exit(1);
  }
  return { appID, appKey };
}

async function searchSchools(state, query, appID, appKey) {
  const params = new URLSearchParams({ st: state, q: query, appID, appKey });
  const res = await fetch(`${API_BASE}/schools?${params}`);
  if (!res.ok) {
    const text = await res.text();
    throw new Error(`API error ${res.status}: ${text}`);
  }
  return res.json();
}

function pad(str, len) {
  return String(str ?? "N/A").slice(0, len).padEnd(len);
}

function displayResults(data) {
  const schools = data.schoolList ?? [];
  const total = data.numberOfSchools ?? 0;

  if (schools.length === 0) {
    console.log("No schools found.");
    return;
  }

  console.log(`\nFound ${total} school(s). Showing first ${schools.length}:\n`);

  const header = `${"School Name".padEnd(40)} ${"City".padEnd(18)} ${"Grades".padEnd(10)} ${"Rank".padEnd(14)} ${"Stars".padEnd(6)} Students`;
  console.log(header);
  console.log("-".repeat(header.length));

  for (const school of schools) {
    const name = pad(school.schoolName, 40);
    const city = pad(school.address?.city, 18);
    const grades = pad(`${school.lowGrade ?? "?"}-${school.highGrade ?? "?"}`, 10);

    // rankStars is on a 0-5 scale; rank history sorted most-recent first
    const r = school.rankHistory?.[0];
    const rank = pad(r ? `${r.rank} of ${r.rankOf}` : "N/A", 14);
    const stars = pad(r?.rankStars, 6);

    const students = school.schoolYearlyDetails?.[0]?.numberOfStudents ?? "N/A";

    console.log(`${name} ${city} ${grades} ${rank} ${stars} ${students}`);
  }
}

async function main() {
  const { appID, appKey } = getCredentials();
  const state = "WA";
  const query = "Lincoln";

  console.log(`Searching for '${query}' schools in ${state}...`);

  try {
    const data = await searchSchools(state, query, appID, appKey);
    displayResults(data);
  } catch (err) {
    console.error(`Error: ${err.message}`);
    process.exit(1);
  }
}

main();
